### PR TITLE
Bug fixes for the reports flow

### LINKED ISSUE #3088: Create JSON for making reports
_This ticket is part of epic #2626_

### Detailed description
This is step 1 in creating scheduled reports. While generating a report, we need to create a JSON.
With every step in the report flow, this JSON needs to be expanded (with a post-request). 
At the end of the flow, the report will be generated based on this JSON. 

This JSON will help us to go back and forth through the generate-report pages.
This will also make it possible to implement other features, like scheduling reports and creating "recipes".

This ticket is pre-work that has to be done before we can implement the scheduled reports.

### Possible solution

The JSON as here described will be "all things needed to make a report". 
This consists of a JSON object that will be passed from view to view with user selections.

What is needed for this JSON?
**- OOIs:** Remember which objects are selected. We don't want to make multiple octopoes requests each time we need the show attributes. 
**- OOIs filters.** We want to remember the filters used for filtering OOIs
- observed_at: When are these OOIs observed when creating the report.
**- Report types:** Which report types are selected?
**- Plugins:** We want to remember the state of the plugins during report creation.

With this information we can generate data for the report types selected: the data together with the JSON we will save in bytes.
The data consists of:
- Calculations to find/fetch vulnerabilities.
- Aggregation of data for aggregate report (the sum of parts) (de-duplication)
- Subreports 

This is an example of how the JSON should look like. If also configs are needed, this can be added to the JSON.
```
report = {
   "observed_at" : "2014-06-13 23:59:59 00:00",
   "ooi_filters": 
      {
           "ooi_types": ["Hostname","URL"]
           "clearance_levels": [2, 3, 4],
           "clearance_types": "["declared", "inherited"]
      },
   "oois" : 
       {
           "pk": "Hostname|internet|mispo.es",
           "clearance_level": 2,
           "clearance_type": "declared"
        },
        {
           "pk": "Hostname|internet|mispo.es",
           "clearance_level": 2,
           "clearance_type": "declared"
        },
    "report_types": ["aggregate_report", dns_report", "findings_report"],
    "plugins": {
        "plugin_id": "dns-records",
        "name": "DnsRecords",
        "description": "Fetch the DNS record(s) of a hostname",
        "type": "boefje",
        "options": "required",
        "scan_level": 2,
      },
       "plugin_id": "dns-records",
        "name": "DnsRecords",
        "description": "Fetch the DNS record(s) of a hostname",
        "type": "boefje",
        "options": "required",
        "scan_level": 2,
        
}
```
### Where do we save report recipes? 
There are some options:
- [x] Create a new Report Recipe OOI and/or add a new OOi Type
- [ ] Report OOI. Save the recipe in a variable so it can be used to generate another report in time. When the OOI is rescheduled we can use the recipe info send to Scheduler to unpack it there. If a report task means, do a rerun for each ooi in the recipe. I am doubting about the plugin information needs to be stored in the recipe. 
image

- [ ] Scheduler. We can also feed the Scheduler the Report OOI with the recipe in it and then extract the data in Scheduler.

- [ ] Report Data. Report Data is stored in bytes. 
image

use also in:
image

To unpack the data for the view.

### LINKED ISSUE #3604: Suggested plugins disappeared from Reports
**Describe the bug**
The suggested plugins section is always empty for generated reports. 

**To Reproduce**

1. Go through the report flow and select DNS. 
2. Make sure you have some plugins enabled.
3. Generate report
4. See empty suggested plugins section. 

**Expected behavior**
The suggested plugins shows which plugins are enabled and disabled for the generated report.

**Screenshots**
image

**OpenKAT version**
main

### LINKED ISSUE #3612: Reports - asset overview empty
**Describe the bug**
When generating a report (e.g. onboarding) the asset overview section (Chapter 1) doesn't contain any data. The plugins section was already identified to be empty before: #3604. 

**To Reproduce**

1. Go through the onboarding for mispo.es
2. Wait until the report is done.
3. Open report and see the empty chapter 1.

**Expected behavior**
Chapter 1 contains an overview of all selected objects, plugins and reports. 

**Screenshots**
image

**OpenKAT version**
main

### PR DESCRIPTION
**Pull request by Rieven**
### Changes

Fix different bugs for reports:
- Findings Report does not have any defined plugins, empty set was fetching all plugins instead of None plugins.
- Report was missing Asset overview data. Fixed to save report data in bytes with the input data.
- 404 when getting Report by id. `observed_at` must be set at end of day when no observed_at is in the URL (from filters, when clearing filters)
- Created with data from: this was showing also the time, while we do not choose time in frontend. If you create a report with the data of today, it will show a inconsistent time end of the day while the created at date is older. Removing the time did the trick.

### Issue link
#3612 
#3604 
#3088 

Closes #3612 
Closes #3604 
Closes #3088 

### Demo
Assets oois and report types
image

plugins
image

Child reports link as report name
image

A Childreport
image

Aggregate report
image

concat with more than 1 ooi
image

concat single ooi
image

from this 
image

To this
image

### QA notes

_Please add some information for QA on how to test the newly created code._

---

### Code Checklist

- [x] All the commits in this PR are properly PGP-signed and verified.
- [x] This PR only contains functionality relevant to the issue.
- [x] I have written unit tests for the changes or fixes I made.
- [x] I have checked the documentation and made changes where necessary.
- [x] I have performed a self-review of my code and refactored it to the best of my abilities.

- [x] Tickets have been created for newly discovered issues.
- [x] For any non-trivial functionality, I have added integration and/or end-to-end tests.
- [x] I have informed others of any required `.env` changes files if required and changed the `.env-dist` accordingly.
- [x] I have included comments in the code to elaborate on what is not self-evident from the code itself, including references to issues and discussions online, or implicit behavior of an interface.

---

## Checklist for code reviewers:

Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_code.md) into your comment.

---

## Checklist for QA:

Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_qa.md) into your comment.

### CODE REVIEW

File: rocky/reports/views/base.py [hunk 1]
```diff
@@ -120,13 +120,24 @@ def get(self, request: HttpRequest, *args: Any, **kwargs: Any) -> HttpResponse:
...
+    merged_plugins = get_plugin_ids(report_types)
+
+    required_plugins_ids = list(merged_plugins["required"])
+    optional_plugins_ids = list(merged_plugins["optional"])
+
+    # avoid empty list getting all plugins from KATalogus
+    if required_plugins_ids:
+        plugins["required"] = sorted(katalogus.get_plugins(ids=required_plugins_ids), key=attrgetter("name"))
+    if optional_plugins_ids:
+        plugins["optional"] = sorted(katalogus.get_plugins(ids=optional_plugins_ids), key=attrgetter("name"))
```
Comment (ammar92) [inline rocky/reports/views/base.py:138]: These are possibly two calls to the list endpoint. You might be better off retrieving all of them in one go and then separate them? E.g. in pseudocode:
```
plugins = get_plugins(ids=required+optional)
required_plugins = filter(lambda x: x in required, plugins)
optional_plugins = filter(lambda x: x in optional, plugins)
...
```

File: rocky/reports/views/base.py [hunk 2]
```diff
@@ -197,6 +210,51 @@ def get_report_type_ids(self) -> list[str]:
...
+        reports_dict = get_report_types_from_aggregate_report(AggregateOrganisationReport)
+        report_types: dict[str, list[dict[str, str]]] = {}
+
+        for option, reports in reports_dict.items():
+            report_types[option] = self.get_report_types_from_ooi_selelection(reports)
+        return report_types
+
+    def get_available_report_types(self) -> tuple[list[dict[str, str]] | dict[str, list[dict[str, str]]], int]:
+        report_types: list[dict[str, str]] | dict[str, list[dict[str, str]]] = {}
+        if self.report_type is not None:
```
Comment (ammar92) [inline rocky/reports/views/base.py:244]: You don't need to check for `None` here if you're not doing anything with it and especially if you're already testing against other conditions

File: rocky/reports/views/base.py [hunk 3]
```diff
@@ -589,8 +609,15 @@ def get(self, request, *args, **kwargs):
 
         return super().get(request, *args, **kwargs)
 
+    def get_observed_at(self):
```
Comment (ammar92) [inline rocky/reports/views/base.py:612]: Can be a property of the view

File: rocky/reports/views/base.py [hunk 4]
```diff
@@ -608,62 +635,88 @@ def get_children_reports(self) -> list[ReportOOI]:
...
+            self.octopoes_api_connector.get(Reference.from_str(ooi), valid_time=self.observed_at) for ooi in ooi_pks
+        ]
+
+    def get_plugins(self, plugins_dict: dict[str, list[str]]) -> list[dict[str, list[Plugin]]]:
+        plugins: dict[str, list[Plugin]] = {"required": [], "optional": []}
+
+        plugin_ids_required = plugins_dict["required"]
+        plugin_ids_optional = plugins_dict["optional"]
+
+        if plugin_ids_required:
```
Comment (ammar92) [inline rocky/reports/views/base.py:649]: See earlier remark about the double request

**Review by ammar92 — COMMENTED** [review]
Nice work, seems like good improvements. I do have a few remarks that could improve it even more

File: rocky/reports/views/base.py (hunk 1 above)
Comment (underdarknl) [reply rocky/reports/views/base.py:138]: Currently, there is no filtered endpoint, and as such this is just a locally cached 'full copy' which we iteratedover twice I'd guess?

Comment (Rieven) [reply rocky/reports/views/base.py:138]: True! Forgot about the calls indeed

**Review by ammar92 — DISMISSED** [review]

**Comment by underdarknl** [comment]
Can we differentiate between the required missing plugins and the optional missing plugins by making the icon red for the missing required plugins?

**Comment by stephanie0x00** [comment]
## Checklist for QA:

- [x] I have checked out this branch, and successfully ran a fresh `make reset`.
- [x] I confirmed that there are no unintended functional regressions in this branch:
  - [x] I have managed to pass the onboarding flow
  - [x] Objects and Findings are created properly
  - [x] Tasks are created and completed properly
- [x] I confirmed that the PR's advertised `feature` or `hotfix` works as intended.
- [x] I checked the logs for errors and/or warnings and made issues where necessary

### What works:
Looks good! All the sections in Chapter 1 are now present again. Reports and aggregate reports can be generated and all subreports can also be opened. 

### What doesn't work:
n/a

### Bug or feature?:
- The rocky logs currently show CSS errors. This doesn't seem to affect the interface experience itself, thus will be picked up in a different ticket. 

image

**Review by underdarknl — APPROVED** [review]

**Pull request merged**